### PR TITLE
Rewrite the Rust generator

### PR DESCRIPTION
**Pull request by hchataing**
- WIP new rust generator
- WIP split generator for root and derived packets
- WIP rewrite serializer
- WIP update generated files

### CODE REVIEW

File: pdl-compiler/src/backends/rust/parser.rs [hunk 1]
```diff
@@ -420,9 +414,11 @@ impl<'a> FieldParser<'a> {
                     // TODO(mgeisler): use
                     // https://doc.rust-lang.org/std/array/fn.try_from_fn.html
                     // when stabilized.
-                    let #id = (0..#count)
-                        .map(|_| #parse_element)
-                        .collect::<Result<Vec<_>, DecodeError>>()?
+                    let mut #id = Vec::with_capacity(#count);
+                    for _n in 0..#count {
```
Comment (mgeisler) [inline pdl-compiler/src/backends/rust/parser.rs:418]: I believe it's more direct to use `_` here: that way you don't get any variable allocated. Right now, I think you're relying on `_n` being optimized away.

```suggestion
                    for _ in 0..#count {
```

File: pdl-compiler/src/backends/rust/parser.rs [hunk 2]
```diff
@@ -463,9 +459,11 @@ impl<'a> FieldParser<'a> {
...
```
Comment (mgeisler) [inline pdl-compiler/src/backends/rust/parser.rs:463]:
```suggestion
                    for _ in 0..#count {
```

File: pdl-compiler/src/backends/rust/serializer.rs [hunk 3]
```diff
@@ -16,131 +16,228 @@ use crate::backends::rust::{mask_bits, types, ToIdent, ToUpperCamelCase};
...
+    fn add_assign(&mut self, other: &RuntimeSize) {
+        self.constant += other.constant;
+        self.variable.extend_from_slice(&other.variable)
+    }
+}
+
+impl quote::ToTokens for RuntimeSize {
+    fn to_tokens(&self, tokens: &mut proc_macro2::TokenStream) {
+        tokens.extend(match self {
+            RuntimeSize { constant, variable } if variable.is_empty() => quote! { #constant },
```
Comment (mgeisler) [inline pdl-compiler/src/backends/rust/serializer.rs:63]: I noticed that the generated code has a few `1usize` constants and I bet they are generated here. It's not really a problem and perhaps it's even nice for understanding that the code reads `123usize + foo.len()` or similar?

If you want to make things consistent, you should be able to slap on a `syn::Index::from(constant)` here and two lines below.

**Review by mgeisler — COMMENTED** [review]
This looks much better than the old code, well done!

File: pdl-compiler/src/backends/rust/serializer.rs (hunk 3 above)
Comment (hchataing) [reply pdl-compiler/src/backends/rust/serializer.rs]: Applied the suggested change :)

**Review by mgeisler — APPROVED** [review]
I think this looks very nice and I imagine the smaller generated files will result in an improvement in code size of the final binaries?

**Comment by hchataing** [comment]
> I think this looks very nice and I imagine the smaller generated files will result in an improvement in code size of the final binaries?

That might be the case, I haven't checked though.
Do you think we should benchmark the parser / serializer ?

**Comment by mgeisler** [comment]
> Do you think we should benchmark the parser / serializer ?

No, I wouldn't bother unless that seems to be a bottleneck for someone?

Of course, setting up a benchmark wouldn't be super hard — perhaps you can try [Divan](https://nikolaivazquez.com/blog/divan/), which I've been meaning to try out myself :smile:

**Pull request merged**